### DISCUSSION TITLE
IPsec SA: only UDP encapsulation is supported

### DISCUSSION THREAD

**Discussion by bin-wang1** [strongswan/strongswan#1755] [Q&A]
friends,I need a help. this is my configuration file：
connections {
    vpn {
        version=1
        proposals = sm4cbc-sm3-sm2dh
        local_addrs=172.12.1.11
        remote_addrs=172.12.1.12
        local {
            auth = pubkey
            id = "172.22.16.2"
            certs = server.cert.pem
        }
        remote {
            auth = pubkey
            id = "172.22.16.4"
        }
        children {
            vpn_child {
                esp_proposals = sm4cbc-sm3-sm2dh
                updown = /opt/ss-gmalg/libexec/ipsec/_updown iptables
            }
        }
    }
}

this is the log:

[IKE] initiating Main Mode IKE_SA vpn[1] to 172.12.1.12
[ENC] generating ID_PROT request 0 [ SA V V V ]
[NET] sending packet: from 172.12.1.11[500] to 172.12.1.12[500] (140 bytes)
[NET] received packet: from 172.12.1.12[500] to 172.12.1.11[500] (140 bytes)
[ENC] parsed ID_PROT response 0 [ SA V V V ]
[IKE] received XAuth vendor ID
[IKE] received DPD vendor ID
[IKE] received FRAGMENTATION vendor ID
[CFG] selected proposal: IKE:SM4_CBC_128/HMAC_SM3/PRF_HMAC_SM3/CURVE_SM2
[ENC] generating ID_PROT request 0 [ KE No ]
[NET] sending packet: from 172.12.1.11[500] to 172.12.1.12[500] (132 bytes)
[NET] received packet: from 172.12.1.12[500] to 172.12.1.11[500] (193 bytes)
[ENC] parsed ID_PROT response 0 [ KE No CERTREQ ]
[IKE] received cert request for 'C=Country, O=Company, CN=root_CA'
[IKE] sending cert request for "C=Country, O=Company, CN=root_CA"
[IKE] authentication of '172.22.16.2' (myself) successful
[IKE] sending end entity cert "C=Country, O=Company, CN=172.22.16.2"
[ENC] generating ID_PROT request 0 [ ID CERT SIG CERTREQ N(INITIAL_CONTACT) ]
[NET] sending packet: from 172.12.1.11[500] to 172.12.1.12[500] (652 bytes)
[NET] received packet: from 172.12.1.12[500] to 172.12.1.11[500] (524 bytes)
[ENC] parsed ID_PROT response 0 [ ID CERT SIG ]
[IKE] received end entity cert "C=Country, O=Company, CN=client"
[CFG]   using certificate "C=Country, O=Company, CN=client"
[CFG]   using trusted ca certificate "C=Country, O=Company, CN=root_CA"
[CFG] checking certificate status of "C=Country, O=Company, CN=client"
[CFG] certificate status is not available
[CFG]   reached self-signed root ca with a path length of 0
[IKE] authentication of '172.22.16.4' with SM2_WITH_SM3 successful
[IKE] IKE_SA vpn[1] established between 172.12.1.11[172.22.16.2]...172.12.1.12[172.22.16.4]
[IKE] scheduling rekeying in 13712s
[IKE] maximum IKE_SA lifetime 15152s
[ENC] generating QUICK_MODE request 3675722089 [ HASH SA No KE ID ID ]
[NET] sending packet: from 172.12.1.11[500] to 172.12.1.12[500] (252 bytes)
[NET] received packet: from 172.12.1.12[500] to 172.12.1.11[500] (252 bytes)
[ENC] parsed QUICK_MODE response 3675722089 [ HASH SA No KE ID ID ]
[CFG] selected proposal: ESP:SM4_CBC_128/HMAC_SM3/CURVE_SM2/NO_EXT_SEQ
[ESP]   IPsec SA: only UDP encapsulation is supported
[ESP] failed to create SAD entry
[ESP]   IPsec SA: only UDP encapsulation is supported
[ESP] failed to create SAD entry
[IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel
initiate failed: establishing CHILD_SA 'vpn_child' failed

**Comment by tobiasbrunner** [comment]
Looks like you are trying to use the [kernel-libipec plugin](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html) with IKEv1. Since the plugin requires UDP encapsulation, by default, it forces that by faking NAT-D hashes. However, your peer doesn't seem to support NAT traversal (NAT-T) or has that feature disabled so UDP encapsulation can't be forced.

You should consider whether you actually need to use the _kernel-libipsec_ plugin or you could disable it and use the _kernel-netlink_ plugin instead, which uses the IPsec stack in the kernel.

Since 5.9.11 it's also possible to enable raw ESP handling for the _kernel-libipsec_ plugin if you decide to go with the userland IPsec implementation.

**Reply by maykolo** [reply]
Hello, I would like to inquire if the national encryption can only be used in NAT-T mode. I have modified various options in the configuration file to disable NAT-T. However, the result is that it still uses UDP NAT mode. By capturing packets, I cannot obtain the ESP protocol, only the UDP protocol. I want to know how to modify it to only the original ESP message because I need to perform packet capture debugging. Thank you.

**Comment by weimengfan** [comment]
您好，这个问题解决了吗，我也在用国密版本的strongswan，用ikev1也是这个问题

**Reply by MJUPCQ** [reply]
哈哈，我刚才试了下。
我用的这个版本git@github.com:highland0971/strongswan-gmalg-merge.git master版本
也出现了这个错误.
然后我尝试修改strongswan/src/libcharon/sa/ikev1/tasks/isakmp_vendor.c文件里的
vendor_natt_ids[]=
{ "NAT-T (RFC 3947)", EXT_NATT, FALSE, TRUE, 16,"\x4a\x13\x1c\x81\x07\x03\x58\x45\x5c\x57\x28\xf2\x0e\x95\x45\x2f"}
修改为
{ "NAT-T (RFC 3947)", EXT_NATT, TRUE, TRUE, 16,"\x4a\x13\x1c\x81\x07\x03\x58\x45\x5c\x57\x28\xf2\x0e\x95\x45\x2f"}
后，就没有这个报错了，并且隧道能够正常使用了。因为目前的报错看起来，是没有正确告知对方我们支持NAT-T,并且我对比了一下之前strongswan5.5.0的版本，这里默认配置的是TRUE，所以我试着主动发送尝试下。但是不清楚为什么新版本不是了，所以后续实际的影响，还需要验证。

**Reply by maykolo** [reply]
Hello, I would like to inquire if the national encryption can only be used in NAT-T mode. I have modified various options in the configuration file to disable NAT-T. However, the result is that it still uses UDP NAT mode. By capturing packets, I cannot obtain the ESP protocol, only the UDP protocol. I want to know how to modify it to only the original ESP message because I need to perform packet capture debugging. Thank you.
